src/sentry/api/endpoints/project_dynamic_sampling.py

```diff
@@ -281,48 +347,128 @@ def get(self, request: Request, project) -> Response:
         )
 
         project_breakdown = None
+        parent_trace_ids = None
+        parent_project_breakdown = []
         if distributed_trace:
             # If the distributedTrace flag was enabled, then we are also interested in fetching
             # the project breakdown of the projects in the trace of the root transaction
-            trace_id_list = [transaction.get("trace") for transaction in root_transactions]
+
+            trace_ids = [transaction.get("trace") for transaction in transactions]
+
             projects_in_org = Project.objects.filter(organization=project.organization).values_list(
                 "id", flat=True
             )
 
-            project_breakdown = self.__run_discover_query(
-                columns=["project_id", "project", "count()"],
-                query=f"event.type:transaction trace:[{','.join(trace_id_list)}]",
-                params={
-                    "start": query_time_range.start_time,
-                    "end": query_time_range.end_time,
-                    "project_id": list(projects_in_org),
-                    "organization_id": project.organization.id,
-                },
-                limit=20,
-                referrer="dynamic-sampling.distribution.fetch-project-breakdown",
-            )
-
+            # Discover query to fetch parent projects (smart query)
+            projects_counts = self.__project_stats_query(project, query_time_range, trace_ids)
             # If the number of the projects in the breakdown is greater than 10 projects,
             # then a question needs to be raised on the eligibility of the org for dynamic sampling
-            if len(project_breakdown) > 10:
+            if len(projects_counts) > 10:
                 return Response(
                     status=status.HTTP_400_BAD_REQUEST,
                     data={
                         "detail": "Way too many projects in the distributed trace's project breakdown"
                     },
                 )
 
+            # Requesting for a project that is root for some transactions,
+            # but not root for others and part of distributed trace like
+            root_projects_count = len(
+                [project for project in projects_counts if project["root"] > 0]
+            )
+            # get requested project id in projects_count dataset
+            requested_project = next(
+                item for item in projects_counts if item["project_id"] == project.id
+            )
+
+            # This condition covers all 3 cases (A, C, D):
+            if root_projects_count == 1:
+                # build parent_project_breakdown
+                parent_project_breakdown = [
+                    {"project_id": project.id, "project": project.slug, "percentage": 1}
+                ]
+            # Case B: the most tricky one and requires extra query
+            elif root_projects_count > 1 and requested_project["root"] > 0:
```

Comment:
```suggestion
            elif root_projects_count > 1:
```
This condition is redundant and actually messes up your logic a bit